File: apps/sdkwork-mcp-flutter-mobile/packages/sdkwork_mcp_flutter_mobile_core/lib/src/services/marketplace_service.dart
import 'package:sdkwork_mcp_app_sdk_generated_flutter/app_client.dart';
import 'package:sdkwork_mcp_app_sdk_generated_flutter/src/models.dart';

class MarketplaceCatalog {
  const MarketplaceCatalog({
    required this.categories,
    required this.servers,
  });

  final List<McpServerCategoryRecord> categories;
  final List<McpServerRecord> servers;
}

class ServerDetailBundle {
  const ServerDetailBundle({
    required this.server,
    required this.tools,
    required this.resources,
    required this.prompts,
  });

  final McpServerRecord server;
  final List<McpToolRecord> tools;
  final List<McpResourceRecord> resources;
  final List<McpPromptRecord> prompts;
}

List<T> _readPageItems<T>(
  dynamic data,
  T Function(Map<String, dynamic> json) fromJson,
) {
  if (data is! Map) {
    return const [];
  }
  final items = data['items'];
  if (items is! List) {
    return const [];
  }
  return items
      .whereType<Map>()
      .map((entry) => fromJson(Map<String, dynamic>.from(entry)))
      .toList(growable: false);
}

McpServerRecord? _readResourceItem(dynamic data) {
  if (data is! Map) {
    return null;
  }
  final item = data['item'];
  if (item is! Map) {
    return null;
  }
  return McpServerRecord.fromJson(Map<String, dynamic>.from(item));
}

const _catalogPageSize = 200;

Future<MarketplaceCatalog> fetchMarketplaceCatalog(SdkworkAppClient client) async {
  final categoriesResponse = await client.mcp.listCategories(null, _catalogPageSize);
  final serversResponse = await client.mcp.listServers(null, _catalogPageSize);
  return MarketplaceCatalog(
    categories: _readPageItems(
      categoriesResponse?.data,
      McpServerCategoryRecord.fromJson,
    ),
    servers: _readPageItems(
      serversResponse?.data,
      McpServerRecord.fromJson,
    ),
  );
}

Future<ServerDetailBundle> fetchServerDetail(
  SdkworkAppClient client,
  String serverKey,
) async {
  final serverResponse = await client.mcp.getServer(serverKey);
  final server = _readResourceItem(serverResponse?.data);
  if (server == null) {
    throw StateError('MCP server not found: $serverKey');
  }
  final serverId = server.id;
  final results = await Future.wait([
    client.mcp.listTools(serverId, null, _catalogPageSize),
    client.mcp.listResources(serverId, null, _catalogPageSize),
    client.mcp.listPrompts(serverId, null, _catalogPageSize),
  ]);
  return ServerDetailBundle(
    server: server,
    tools: _readPageItems(results[0]?.data, McpToolRecord.fromJson),
    resources: _readPageItems(results[1]?.data, McpResourceRecord.fromJson),
    prompts: _readPageItems(results[2]?.data, McpPromptRecord.fromJson),
  );
}

File: apps/sdkwork-mcp-h5/packages/sdkwork-mcp-h5-core/src/services/marketplaceService.ts
import type {
  McpPromptRecord,
  McpResourceRecord,
  McpServerCategoryRecord,
  McpServerRecord,
  McpToolRecord,
} from 'sdkwork-mcp-app-sdk-generated-typescript/src/types';

import { getMCPAppSdkClientWithSession } from '../sdk/mcpAppSdkClient';
import { unwrapSdkWorkPage } from '../sdk/sdkPage';

const catalogListParams = { pageSize: 200 } as const;

export async function fetchMarketplaceCatalog() {
  const client = getMCPAppSdkClientWithSession();
  const [categoryResponse, serverResponse] = await Promise.all([
    client.mcp.listCategories(catalogListParams),
    client.mcp.listServers(catalogListParams),
  ]);
  return {
    categories: unwrapSdkWorkPage<McpServerCategoryRecord>(categoryResponse).items,
    servers: unwrapSdkWorkPage<McpServerRecord>(serverResponse).items,
  };
}

export async function fetchServerDetail(serverKey: string) {
  const client = getMCPAppSdkClientWithSession();
  const server = await client.mcp.getServer(serverKey);
  const serverId = String(server.id);
  const [tools, resources, prompts] = await Promise.all([
    client.mcp.listTools(serverId, catalogListParams),
    client.mcp.listResources(serverId, catalogListParams),
    client.mcp.listPrompts(serverId, catalogListParams),
  ]);
  return {
    server,
    tools: unwrapSdkWorkPage<McpToolRecord>(tools).items,
    resources: unwrapSdkWorkPage<McpResourceRecord>(resources).items,
    prompts: unwrapSdkWorkPage<McpPromptRecord>(prompts).items,
  };
}
